### ISSUE TITLE
Menu "Analytic" - scenarios (Part 1)

### ISSUE THREAD

**Issue by dhyougit** [amitprasad531/gpw13-dashboard#4]
@amitprasad531 

Brainstorming for the "Analytic" menu that is not very heavy on the visualization side of work. Part 1 is more focused on the analytic side of suggestion. (Part 2 will be more on simple analytics that is focused on visualization.)

**A, (What are you looking for?) Natural language based search engine** 
- Visualization: simple search bar like a google front page
- Purpose and benefit: Present a simple search bar that users can type their questions in plain language and show the most relevant data from the database.  This won't be much of fancy NLP however just a topic modeling based on tags that we can generate per indicator or use the indicator name itself. 

**B. (How it is connected to other indicators?) ** 
Option A. Heatmap
- Visualization: Heatmap
Option B. Dendogram 
- Visualization: Dendogram

- Purpose and benefit:  Show the connection to other indicators within the dataset. 

**C. (Whom they are connected with?) Clustering: (Yes, again - !)** 
- Visualization: Boxplots per cluster or plain x(UHC)-y (HALE) scatter plot of countries
- Purpose and benefit: K-mean clustering to show which countries are sharing a similar situation. To make it simple, I suggest to use the HALE and UHC coverage index only and show the 4-5 clusters of countries 

**D. (How is the world/your interest of the country will be like in 2023?) Forecasting**
- Visualization: Single indicator (Allow selection of continuous variable format indicator or just show HALE) trend forecasting 
- Purpose and benefit: Show past trend and forecast 2023. But 95% CI can be provided as a default option (..boring) so additionally, give propensity score of the prediction to show the accuracy of prediction. 

-

**Comment by amitprasad531** [comment]
Comments on Part 1:

A. Excellent idea but I think it might be better suited to search for data through tables

B. Heatmaps are fairly popular, dendograms less so. Since this is an app for a general audience I would focus on more complex applications later. Two points here. 

- The analytics can have several sub menus
- Heatmaps could be a special visualization we provide as a sub menu item
- Dendograms for later unless they are easy to render

C. I would see the use of boxplots and clustering as part of the core analytics. These are great ideas and we should expand on them.

D. Forecasting can be done in two forms:
- Simple (linear)
- Regression-based model
We should use both options. This would be another sub-menu under analytics.

An extra comment: given this plan I will drop the "UHC" menu tab.

**Comment by dhyougit** [comment]
Branched code merged which will plot below two graphs in the analytics tab. It is not fully interactive graph yet because I have to make the selected item into a global variable or change the server.R file. 


![dendrogram](https://user-images.githubusercontent.com/21180571/53260328-82152780-36d1-11e9-895e-4753eb059980.png)
![heatmap](https://user-images.githubusercontent.com/21180571/53260337-86d9db80-36d1-11e9-94bc-dd1a30d88a2b.png)

**Comment by amitprasad531** [comment]
Hi Doohee,

Great work! These look amazing and if we are able to give people the
flexibility to make their own charts it will make a big difference.

Can you also label everything clearly so that anyone can figure out what
messages to take away from these graphs?

Thanks, Amit.

**Comment by amitprasad531** [comment]
I have merged your branch with the master. Now your code is included in the
master branch.

Take a look at the dashboard I have created over the last two days. I have
3 major tasks to complete the first version:
1. Prepare the data file from all the files that you downloaded on Wednesday
2. Add the ability to select GPW indicators to display on the dashboard
3. Get the download button to work! :)

Enjoy your weekend.

**Comment by dhyougit** [comment]
Awesome and beautiful dashboard! The first page itself is already enough information for MSs! Thanks for relentless innovation effort on this app. 

Everything work impeccable on my Mac environment as well except the dendrogram graph I built. So I am building a more regional level of analysis instead of country level. I will update the code once it is ready. 

Git commands finally worked in my Mac environment more smoothly and I managed to commit some changes as below. 
![screen shot 2019-02-23 at 10 14 27 pm](https://user-images.githubusercontent.com/21180571/53291989-368b7800-37bc-11e9-84f6-5c85016e8bd2.png)
The updates are not reflected in the repo (maybe there is a bit of lag) though I merged into branch name under GHObasedWHS_3 tree. ;) 

Bon weekend!

**Comment by amitprasad531** [comment]
Thanks Doohee! I think we should accelerate work on this app and get a good
draft ready for 'beta' review by mid-March. Somnath can join us in the
alpha review - we can show it to him next week (not the coming one).

I have just created the mega dataset (sdg.csv in the main data folder). It
has data for 194 countries, 28 years, and 102 columns which is more than
half a million table cells. It includes all the SDGs but more than that.

Next step is to use this dataset in the main app and increase flexibility
of use. Let me know once you have committed your changes and I can merge
them with the master. Ideally we should sit together to resolve any
conflicts with the files.

Anyway, see you tomorrow.

**Comment by dhyougit** [comment]
Thanks for the amazing data prep work! It would take so long if I did but you made it so quickly! 

To your point, we can do alpha testing with Somnath by next week. 
I built a structure to make an interactive model builder in the forecast tab. I would love to merge the code on githut but somehow the WHO network does not like github push/pull commands. Anyhow, the outcome graph will look very similar as what you saw from my previous kmean classification based on the UHC (or we can select to IHR) index based on each countries' complete list of indicators. 
I found some little errors to fix but would like to share what's on the way. But if you found it interesting, please feel free to fix it! ;) 

I will upload the file in separate filename so that it does not mess up other files.

**Comment by amitprasad531** [comment]
Fantastic progress on the analytics Doohee! I hope to join you next week and take on the analytics section. It should be the most exciting part of the app to work and I’m going to rely on many of your ideas for new and interesting analysis that people can do with data. Keep working on it and hopefully we can merge the app bits soon.

**Comment by dhyougit** [comment]
@amitprasad531 
HELP !!!!!!!!!!!!! 
Uploaded ui.r file work fine and can be plugged into a the master code branch. 
Below part of code is for server.r, which works fine in outside by itself, but it returns an error when it runs inside of the shiny server code. I think I messed up with the reactive term and been stucked for two days.. Could you have a quick look please???? 


`library(cluster)  

# remove NAs 
df_pre0<-reactive({whs%>%filter(year==input$year)}) # filter data to selected year 
row.names(df_pre0) <- df_pre0()$country # give country name as index 
df_pre1<-within(df_pre0, rm(country,year,iso3, Region, mort5_14)) # make df narrow --- mort5_14: format mismatch 
df_pre2<-df_pre1[, -which(colMeans(is.na(df_pre1)) > 0.25)] #make df shorter 
df_pre2<- na.omit(df_pre2) # drop NAs
#df_pre2<-reactive({df_pre2[, c(input$xcol)]}) #select input variables from the clead DF # IMPORTANT : THIS SHOULD BE AT LEAST MORE THAN 5 INDICATORS 
coltodrop<-reactive({input$ycol}) 
df_pre3<-within(df_pre2(), rm(coltodrop)) # drop output var to make DF scaleable 
# Standardize 
df_pre3 <-scale(df_pre3)

#kmean
set.seed(100)
centernum<-reactive({input$clusters})
km<-kmeans(df_pre3, centers = centernum, nstart = 10)
df_out<-km$cluster
df_out<-cbind(df_out, df_pre2()$coltodrop)
colnames(df_out) <- c("label", "ycol")

# violinplot 
df_out<-as.data.frame(df_out)
df_out$label <- as.factor(df_out$label)
ggplot(df_out(), aes(x=label, y=ycol, fill=label)) + 
  geom_violin()+ geom_jitter(shape=16, position=position_jitter(0.2))`

**Comment by amitprasad531** [comment]
I can see some errors. Send me your files by email. Do not commit to Github.
